EIPS/eip-7484.md

```diff
@@ -0,0 +1,239 @@
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Definitions
+
+- **Smart account** - An ERC-4337 compliant smart contract account that has a modular architecture.
+- **Module** - Self-contained smart account functionality.
+- **Attestation** - Onchain assertions made about the security of a module.
+- **Registry** - An onchain list of attestations about modules.
+- **Adapter** - Smart account functionality that handles the fetching and validation of attestations from the registry.
+- **Attester** - An entity that makes an attestation about a module.
+
+### Overview
+
+![Adapter flow](../assets/eip-7484/flow.png)
+
+An Adapter is a piece of functionality that allows smart accounts to query the module registry before using a module. In order to be compatible with this proposal, a smart account MUST implement the adapter either natively in the account or as a default module that is installed during account deployment.
+
+The adapter:
+
+- MUST query the registry at least once before calling a module for the first time.
+- MUST use one of the functions on the IRegistry interface when querying the registry.
+- Is RECOMMENDED to query the registry before module installation.
+- MAY query the registry before module execution.
+
+### Interfaces
+
+#### IRegistry.sol
+
+##### `check`
+
+Checks the attestation of a given module address and attestations made by a single attester.
+
+Returns two timestamps. `listedAt` is a `uint48` timestamp of when the attestation was made.
+The second return parameter `revokedAt` is a `uint48` timestamp of when an attestation was revoked.
+
+**NOTE**:
+
+- `listedAt` of 0 MUST be treated as a security risk.
+- `revokedAt` of non-0 MUST be treated as a security risk.
+
+```solidity
+function check(
+        address module,
+        address attester
+    )
+        public
+        view
+        returns (uint48 listedAt, uint48 revokedAt);
+```
+
+##### `verify`
+
+Verifies the attestations of a given module and multiple attesters. Additionally a threshold can be provided.
+
+Returns a boolean value `verified` if threshold `>=` the number of attesters that attested to this module without revoking the attestation.
+
+_EXAMPLE_: If threshold of 2 is set and 5 attesters are queried, the function will return `true` if 2 or more of the selected attesters have issued an attestation that has not been revoked.
+
+**NOTE**:
+
+- `verified` of false MUST be treated as a security risk.
+
+```solidity
+function verify(
+        address module,
+        address[] memory attesters,
```

Comment:
Are there any gas limits involved when iterating the calls for the attestations?

I.e. if the second `attester` uses up all gas, would this block any further execution? 

Should the caller make any assumptions here (or set any gas limit)?

Reply:
I'm not entirely sure what you mean. In the reference implementation, neither `check` nor `verify` read the attestation data so I'm unsure how the call could run out of gas (unless the gas limit was not properly set on the frontend). 

If you refer to cases in which data is being read and excessively long attestations could cause the wallet to run out of gas then I'm not sure whether that's a feasible attack vector since creating these attestations would be quite expensive.